### ISSUE TITLE
Volume defined using volumes_from is not updated when stopping, recreating and starting the source service

### ISSUE THREAD

**Issue by jaapz** [docker/compose#3329]
I have defined volumes in a compose file like this:

```
version: '2'
services:
    nginx:
        ...
        volumes_from:
            - web

    web:
        ...
        volumes:
            - /var/www/myapp/static
```

The nginx container is configured to serve static files from the volume it gets from the web container. All other requests are proxied to the app running in the web container.

Getting things up and running using `docker-compose up` works just fine. However the problem starts when I restart the running web containers.

To have as little downtime as possible, I want to update the running web containers to the latest versions of my image, and recreate and restart them without having to do `docker-compose down` and `docker-compose up`. So I do:

```
$ docker-compose pull web
$ docker-compose stop web
$ docker-compose rm -f web
$ docker-compose start web
```

This makes sure only the web containers are downed, and the nginx container can keep running and accepting requests and proxy requests to other containers.

And even though static files in the defined volumes have changed, they are not updated in the nginx container.

Basically what I need is the volumes defined in volumes_from to be re-attached or re-created when the source container is recreated. Is that possible?
## Versions

```
$ docker --version
Docker version 1.10.3, build 20f81dd
$ docker-compose --version
docker-compose version 1.6.2, build 4d72027
```

**Comment by mkuzmin** [comment]
Also reported in #2064.

**Comment by jaapz** [comment]
Is that because

    volumes_from:
            - web

maybe expands to this?

    volumes_from:
            - web_1

**Comment by migueleliasweb** [comment]
I'm having the same problem with similar versions of docker-compose and docker daemon.

I did some debugging via `docker events` and found that compose actually _somehow_ reuses the volumes as mentioned on #1981 . I certainly don't agree much with the decision but what I did was instead of mounting the static folder directly through the docker-compose.yml I mounted a `/static` folder and inside my entrypoint, right after I generate all static files inside the `/var/www/myapp/static` folder, I move them to the `/static` folder and they get updated as expected.

It's a kida dirty workaround, but it works....for while...in production...

**Comment by jaapz** [comment]
@migueleliasweb that means you generate the files when you start the container, yes?

**Comment by migueleliasweb** [comment]
@jaapz No, I just move then to the volume after starting the container.

**Comment by jaapz** [comment]
Ah, I'll try that workaround, thanks

**Comment by alexist** [comment]
Any plan to add an option to docker-compose in order to avoid this workaround ?

**Comment by aanand** [comment]
volumes_from is almost never a good idea. Use a named volume instead:

```yaml
version: '2'
services:
    nginx:
        ...
        volumes:
            - static:/var/www/myapp/static

    web:
        ...
        volumes:
            - static:/var/www/myapp/static
volumes:
    static: {}
```

**Comment by alexist** [comment]
We are using volumes-from to "inject" configuration into some container. We are building generic image that can be deployed as container in any environment (dev, acceptance, production) without any modification. Thoses images contains a java application server and our webapp. 

To manage the configuration specific to each environment (some external webservice urls for example), we are building configurations images : conf-dev, conf-acceptance, conf-production;

Configuration are loaded into application container thanks to volumes_from.

**Comment by aanand** [comment]
@alexist What's preventing you from using a named volume for this use case?

**Comment by alexist** [comment]
When the configuration image is updated, docker compose recreate the configuration container, but the application container is not updated. The application container don't see the new configuration. We have to manually execute down + up in order to have the new configuration available in the application container.

**Comment by stefda** [comment]
Same here. Specifically, I have a `Dockerfile` that exposes a volume:

```
...
VOLUME /var/www/api
```

and a `docker-compose.yml` that brings it all together:

```
symfony:
  build: ./symfony

fpm:
  build: ./fpm
  volumes_from:
    - symfony
  ports:
    - "9000:9000"

nginx:
  build: ./nginx
  volumes_from:
    - symfony
  ports:
    - "80:80"
    - "443:443"
  links:
    - fpm
```

However, after updating any of the files in the volume container the nginx and fpm containers see the "old" version of those files.

What I'm expecting is that whenever I update files in the volume container and recreate it, the changes will be reflected in the conatiners connected to it without recreating them.

**Comment by amir20** [comment]
Was there any good solution to this? Is using a named volume best option as @aanand suggests?

**Comment by amir20** [comment]
@aanand doesn't work me still. When I update one data container, the named volume doesn't update to latest data 👎

**Comment by rudyryk** [comment]
@aanand Seems like named volumes doesn't solve the original problem. Volumes content is still not updated after image updates.

And actually this is documented behaviour as I can see from [here] (https://docs.docker.com/engine/tutorials/dockervolumes/) and it comes from Docker itself, not compose tool:

- Volumes are initialized when a container is created.
- ...
- **Changes to a data volume will not be included when you update an image.**

This is confusing but it just works like that.

**Comment by rudyryk** [comment]
I finally ended up copying static web content to web image on *every rebuild*.

Another option is to have an extra shared mounted volume to which we collect web content on *every restart*. It works well for example with Django, it has `manage.py collectstatic` script which copies static files from multiple places to single directory.

**Comment by euri16** [comment]
@rudyryk, @amir20 can you please show me how you did it?

**Comment by amir20** [comment]
@euri16 how far did you get? I ended up ding `rsync` which worked perfectly.

**Comment by Nagoo** [comment]
Any news around this issue?. We are facing the same issue with both volumes_from & named volumes. They don't get updated at service update.

**Comment by jaapz** [comment]
Was never fixed for me, I just build docker images containing static files for that release now. In our situation that's a better solution anyway.

**Comment by dchersey** [comment]
See my latest comment on #1981 -- I found a fairly workable  solution using `docker volume rm` as part of my update script.
